### PR TITLE
fix(constraints): make constraints --fix persist changes to manifest

### LINKED ISSUE #2242: [Bug] `yarn constraints --fix` silently does not fix reported violations
- [ ] I'd be willing to implement a fix

**Describe the bug**

Running `yarn constraints --fix` does not fix the issues that are reported with `yarn constraints` and does not issue a warning / error either.

**To Reproduce**

Use this constraints.pro file
```
gen_enforced_dependency(WorkspaceCwd, 'typescript', DependencyRange2, DependencyType) :-
  workspace_has_dependency(WorkspaceCwd, 'typescript', DependencyRange, DependencyType),
  workspace_has_dependency('.', 'typescript', DependencyRange2, DependencyType2),
  DependencyRange \= DependencyRange2.
```

Install a different version of typescript in a workspace than top level.

```
> yarn constraints 
➤ YN0024: workspace must depend on typescript via 3.8, but uses ^4.1.3 instead (in dependencies)
➤ YN0000: Failed with errors in 0s 16ms
> yarn constraints --fix
➤ YN0000: Done in 0s 19ms
```

<details>
  <summary>Reproduction</summary>

(I could not get this fully to work, but maybe this helps?)

  ```js repro
const {promises: {writeFile}} = require("fs")

await packageJson({
  "dependencies": {
    "typescript": "4.1"
  }
}, { cwd: "workspace" })

await packageJsonAndInstall({
  "workspaces": [
    "workspace"
  ],
  "dependencies": {
    "typescript": "3.8"
  }
})

await writeFile("constraints.pro", `
gen_enforced_dependency(WorkspaceCwd, 'typescript', DependencyRange2, DependencyType) :-
  workspace_has_dependency(WorkspaceCwd, 'typescript', DependencyRange, DependencyType),
  workspace_has_dependency('.', 'typescript', DependencyRange2, DependencyType2),
  DependencyRange \= DependencyRange2.
`)

await yarn('plugin', 'import', 'constraints')

await expect(yarn('constraints')).toContain("workspace must depend on typescript via 3.8")

await expect(yarn('constraints', '--fix')).toContain("\n") // multiple lines
  ```
</details>

### PR DESCRIPTION
**Pull request by bgotink**
**What's the problem this PR addresses?**

The change introduced in #1775 to fix fixing the field constraints
broke fixing dependency constraints, by overriding the changes made
to the manifest before persisting.

Fixes #2242

**How did you fix it?**

Dependency constraints work on the Manifest, field constraints work on the raw JSON object. Sync the two after every execution of the constraints to ensure every iteration uses the correct values.

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

**Review by merceyz — CHANGES_REQUESTED** [review]
We should have some tests to ensure we don't regress on this in the future (granted I should have added some for #1775 but oh well)

File: packages/acceptance-tests/pkg-tests-specs/sources/commands/constraints/fix.test.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+const {
+  fs: {writeFile, writeJson, readJson},
+} = require(`pkg-tests-core`);
```
[the remark this answers is not in the document]
Comment (bgotink) [reply packages/acceptance-tests/pkg-tests-specs/sources/commands/constraints/fix.test.js]: I've replaced these with xfs

I copied the `constraints` test file to start from, and I'm sure a lot of tests are created by copying an existing test file. If existing tests aren't updated to use xfs, I'd expect a lot of these kind of comments to pop up.

There are even test files that use both xfs and this `fs` import in the same test.

**Review by merceyz — APPROVED** [review]
Looks good :+1:

**Comment by merceyz** [comment]
This seems to introduce a different bug, when a fix is applied the content of array fields are duplicated

**Pull request merged**